### PR TITLE
Update ippool controller

### PR DESCRIPTION
**Pull request by zhengxiexie**
Update type of ippool to optional.
Overwrite DefaultSubnetAccessMode of vpcnetworkconfig to ippool's type.
Get private or public IpBlockPath from VPCInfo.
Update Type public/private to Public/Private.
Try next ipblock if the previous are exhausted
	If the error message contains `not have spare capacity to satisfy`,
	it shows the ip block has no more capacity, then try next ip block
	from the list of PrivateIpv4Blocks or ExternalIpv4Blocks
	to build IpBlockPath.

### CODE REVIEW

File: pkg/controllers/ippool/ippool_controller.go
Comment (dantingl) [inline pkg/controllers/ippool/ippool_controller.go:135]: prefixLength should also be got from default subnet size?

Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:135]: > prefixLength should also be got from default subnet size?
Actutally, PrefixLength is defined in SubnetRequest, it is specified by WCP.

File: pkg/controllers/ippool/ippool_controller_test.go
Comment (dantingl) [inline pkg/controllers/ippool/ippool_controller_test.go:189]: Public

File: pkg/nsx/services/ippool/builder.go
Comment (dantingl) [inline pkg/nsx/services/ippool/builder.go:103]: For IP exhausted case, we need to return an available ip block instead of the first one

Comment (lxiaopei) [reply pkg/nsx/services/ippool/builder.go:103]: agree

Comment (zhengxiexie) [reply pkg/nsx/services/ippool/builder.go:103]: would fix it after tao's patch merged.

File: pkg/nsx/services/ippool/builder.go
Comment (zhengxiexie) [reply pkg/nsx/services/ippool/builder.go:105]: would fix it after tao's patch merged.

File: pkg/nsx/services/ippool/builder.go
Comment (dantingl) [reply pkg/nsx/services/ippool/builder.go:103]: Add TODO for this one

File: pkg/nsx/services/ippool/builder.go
Comment (dantingl) [inline pkg/nsx/services/ippool/builder.go:95]: In line 94, what is SubnetRequest from v1alpha2?

Comment (zhengxiexie) [reply pkg/nsx/services/ippool/builder.go:95]: `// SubnetRequest defines the subnet allocation request.
type SubnetRequest struct {
	// PrefixLength defines prefix length for this subnet.
	PrefixLength int `json:"prefixLength,omitempty"`

	// IPFamily defines the IP family type for this subnet, could be IPv4 or IPv6.
	// This is optional, the default is IPv4.
	// +kubebuilder:validation:Enum=IPv4;IPv6
	// +kubebuilder:default=IPv4
	IPFamily string `json:"ipFamily,omitempty"`

	// Name defines the name of this subnet.
	Name string `json:"name"`
}`

**Comment by timdengyun** [comment]
please don't merge tao's patch into your patch: Dump error return from SDK to ApiError,
you can use it for your test, but should not be merged in your patch.

Also, please squash 3 commits into one commit.

File: pkg/controllers/ippool/ippool_controller.go
Comment (TaoZou1) [inline pkg/controllers/ippool/ippool_controller.go:144]: why use log.Error but error is nil?

File: pkg/nsx/services/ippool/ippool.go
Comment (TaoZou1) [inline pkg/nsx/services/ippool/ippool.go:78]: if first  *ipSubnet.IpBlockPath == "" , does that means all ipblock are exhausted? 
should we iterate all ipsubnet?

File: pkg/controllers/ippool/ippool_controller.go
Comment (dantingl) [reply pkg/controllers/ippool/ippool_controller.go:144]: This is the way when we do not have error defined but want to print error message.
But I also saw Xie defined err in line 143, so I guess we can just leverage it?

File: pkg/controllers/ippool/ippool_controller.go
Comment (dantingl) [inline pkg/controllers/ippool/ippool_controller.go:157]: Let's do not compare error message, NSX may change error message, error code is good enough

File: pkg/controllers/ippool/ippool_controller.go
Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:144]: ok

File: pkg/nsx/services/ippool/ippool.go
Comment (zhengxiexie) [reply pkg/nsx/services/ippool/ippool.go:78]: no need, if *ipSubnet.IpBlockPath == "", it represents all exhausted, it is the logic indeed.

File: pkg/controllers/ippool/ippool_controller.go
Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:157]: I'm not sure 520012 is only for exhausted, so add an error message check to insure more safety.

Comment (dantingl) [reply pkg/controllers/ippool/ippool_controller.go:157]: Yes. 520012 is only for exhaustion case

Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:157]: OK

File: pkg/controllers/ippool/ippool_controller.go
Comment (TaoZou1) [inline pkg/controllers/ippool/ippool_controller.go:175]: add a log.Error before turn?

Comment (zhengxiexie) [reply pkg/controllers/ippool/ippool_controller.go:176]: ok

**Review by TaoZou1 — APPROVED** [review]

**Pull request merged**
